Don't run torch.compile under non-strict export.

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #165322


While non-strict tracing, we could encounter torch.compile region and the expected behaviour is we don't run torch.compile at all. There are some legacy behaviour for other HOPs like cond, while loop to still run torch.compile under non-strict. For now we still allow them for selected HOPs (cond, while_loop, invoke_subgraph, flex_attention, and strict_mode) For all other use cases, we will turn off torch.compile under non-strict. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @chenyang78

tugsbayasgalan committed:
More robust is_exporting check for compile

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "More robust is_exporting check for compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "More robust is_exporting check for compile"



While non-strict tracing, we could encounter torch.compile region and the expected behaviour is we still run torch.compile but with eager backend because any other backend wouldn't make sense in export context. Previously we apply this special inside explicit torch.compile API, but this seems not enough when we initialize torch.compile outside of export region and later it gets recompiled. This PR fixes this by adding another check inside more lower level API. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "More robust is_exporting check for compile"



While non-strict tracing, we could encounter torch.compile region and the expected behaviour is we still run torch.compile but with eager backend because any other backend wouldn't make sense in export context. Previously we apply this special inside explicit torch.compile API, but this seems not enough when we initialize torch.compile outside of export region and later it gets recompiled. This PR fixes this by adding another check inside more lower level API. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "More robust is_exporting check for compile"



While non-strict tracing, we could encounter torch.compile region and the expected behaviour is we still run torch.compile but with eager backend because any other backend wouldn't make sense in export context. Previously we apply this special inside explicit torch.compile API, but this seems not enough when we initialize torch.compile outside of export region and later it gets recompiled. This PR fixes this by adding another check inside more lower level API. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

ezyang reviewed (approved):
The new code is simpler, so I'm wiling to give it a spin, but I don't really understand the global implications

tugsbayasgalan commented:
@pytorchbot merge

tugsbayasgalan committed:
Update on "Don't run torch.compile under non-strict export. "



While non-strict tracing, we could encounter torch.compile region and the expected behaviour is we don't run torch.compile at all. There are some legacy behaviour for other HOPs like cond, while loop to still run torch.compile under non-strict. For now we still allow them for selected HOPs (cond, while_loop, invoke_subgraph, flex_attention, and strict_mode) For all other use cases, we will turn off torch.compile under non-strict. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo chenyang78

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Don't run torch.compile under non-strict export. "



While non-strict tracing, we could encounter torch.compile region and the expected behaviour is we don't run torch.compile at all. There are some legacy behaviour for other HOPs like cond, while loop to still run torch.compile under non-strict. For now we still allow them for selected HOPs (cond, while_loop, invoke_subgraph, flex_attention, and strict_mode) For all other use cases, we will turn off torch.compile under non-strict. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo chenyang78

[ghstack-poisoned]

ydwu4 reviewed (approved):
Thanks for cleaning up the backend wrapper logic! Turning off torch.compile in non-strict regions for other cases can simplify things a lot.
